platform-sdk/swirlds-merkledb/src/main/java/com/swirlds/merkledb/files/hashmap/HalfDiskHashMap.java

```diff
@@ -425,93 +456,198 @@ public DataFileReader endWriting() throws IOException {
+                        oneTransactionsData.keyValuesView().iterator();
+                fileCollection.startWriting();
+                resetEndWriting(size);
+                final ForkJoinPool pool = getFlushingPool();
+                // Create all the tasks and submit them to the fork-join pool. Number of tasks
+                // running in parallel is limited by the size of the pool. Total number of
+                // tasks created is limited by the size of oneTransactionsData, which is not
+                // expected to be huge, since data flushes are limited
+                while (it.hasNext()) {
+                    IntObjectPair<BucketMutation> keyValue = it.next();
                     final int bucketIndex = keyValue.getOne();
                     final BucketMutation bucketMap = keyValue.getTwo();
-                    flushExecutor.execute(() -> readUpdateQueueBucket(bucketIndex, bucketMap, queue));
-                    ++inFlight;
+                    // Create a "read bucket" task. Executed, once its out is set
+                    final ReadBucketTask readBucketTask = new ReadBucketTask(pool, bucketIndex);
+                    // Create an "update bucket" task. It's the out for the previous task. Executed
+                    // after the bucket is read by the read task and set as input for this task
+                    final UpdateBucketTask updateBucketTask = new UpdateBucketTask(pool, bucketMap);
+                    // This will trigger read bucket task execution
+                    readBucketTask.setOut(updateBucketTask);
                 }
-
-                ReadBucketResult res;
-                while ((res = queue.poll()) != null) {
-                    --inFlight;
-                    if (res.error != null) {
-                        throw new RuntimeException(res.error);
-                    }
-                    try (final Bucket bucket = res.bucket) {
-                        final int bucketIndex = bucket.getBucketIndex();
-                        if (bucket.isEmpty()) {
-                            // bucket is missing or empty, remove it from the index
-                            bucketIndexToBucketLocation.remove(bucketIndex);
-                        } else {
-                            // save bucket
-                            final long bucketLocation =
-                                    fileCollection.storeDataItem(bucket::writeTo, bucket.sizeInBytes());
-                            // update bucketIndexToBucketLocation
-                            bucketIndexToBucketLocation.put(bucketIndex, bucketLocation);
-                        }
-                    } finally {
-                        ++processed;
-                    }
+                // Wait until all tasks are completed. The latch is set in the end of the last
+                // executed "store bucket" task or when an exception is occurred
+                finishWritingLatch.await();
```

Comment:
Minor: If the iterator above is empty, then the latch may never be released. And if the tasks block for whatever reason, it also may not be released. Should this `await()` call use a timeout and error out if one occurs?

Reply:
Correct, this is why this code is only run if `size > 0`, there is a check about it above

Reply:
Sounds good.